core/src/avm2/bytearray.rs

```diff
@@ -33,272 +65,162 @@ impl ByteArrayStorage {
         ByteArrayStorage {
             bytes: Vec::new(),
-            position: 0,
+            position: Cell::new(0),
             endian: Endian::Big,
         }
     }
 
-    /// Write a byte at next position in the bytearray
-    pub fn write_byte(&mut self, byte: u8) {
-        let bytes_len = self.bytes.len();
-        // Allocate space for the byte
-        self.position += 1;
-        if self.position > bytes_len {
-            self.bytes.resize(self.position, 0);
-        }
-        self.bytes[self.position - 1] = byte;
-    }
-
-    /// Write bytes at next position in bytearray (This function is similar to whats in std::io::Cursor)
-    pub fn write_bytes(&mut self, buf: &[u8]) {
-        // Make sure the internal buffer is as least as big as where we
-        // currently are
-        let len = self.bytes.len();
-        if len < self.position {
-            // use `resize` so that the zero filling is as efficient as possible
-            self.bytes.resize(self.position, 0);
-        }
-        // Figure out what bytes will be used to overwrite what's currently
-        // there (left), and what will be appended on the end (right)
-        {
-            let space = self.bytes.len() - self.position;
-            let (left, right) = buf.split_at(cmp::min(space, buf.len()));
-            self.bytes[self.position..self.position + left.len()].copy_from_slice(left);
-            self.bytes.extend_from_slice(right);
-        }
-
-        // Bump us forward
-        self.position += buf.len();
+    /// Write bytes at the next position in the ByteArray, growing if needed.
+    #[inline]
+    pub fn write_bytes(&mut self, buf: &[u8]) -> Result<(), Error> {
+        self.write_at(buf, self.position.get())?;
+        self.position.set(self.position.get() + buf.len());
+        Ok(())
     }
 
-    // Write bytes at an offset, ignoring the current position
-    pub fn write_bytes_at(&mut self, buf: &[u8], offset: usize) {
-        // Make sure the internal buffer is as least as big as where we
-        // currently are
-        let len = self.bytes.len();
-        if len < offset {
-            // use `resize` so that the zero filling is as efficient as possible
-            self.bytes.resize(offset, 0);
+    /// Reads any amount of bytes from the current position in the ByteArray
+    #[inline]
+    pub fn read_bytes(&self, amnt: usize) -> Result<&[u8], Error> {
+        let bytes = self.read_at(amnt, self.position.get())?;
+        self.position.set(self.position.get() + amnt);
+        Ok(bytes)
+    }
+
+    /// Reads any amount of bytes at any offset in the ByteArray
+    #[inline]
+    pub fn read_at(&self, amnt: usize, offset: usize) -> Result<&[u8], Error> {
+        self.bytes
+            .get(
+                offset
+                    ..offset
+                        .checked_add(amnt)
+                        .ok_or("RangeError: Cannot overflow usize")?,
+            )
+            .ok_or_else(|| "EOFError: Reached EOF".into())
```

Comment:
A nice trick for this kind of thing is to do `slice[start..][..len]`. Something like:
```
        self.bytes
            .get(offset..)
            .and_then(|bytes| bytes.get(..amnt))
```